presto-orc/src/test/java/com/facebook/presto/orc/IntegerDictionaryProviderTest.java

```diff
@@ -0,0 +1,302 @@
+            throws IOException
+    {
+        NodeId nodeId = new NodeId(1, 0);
+        long[] data = new long[] {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16};
+        IntegerDictionaryProvider dictionaryProvider = new IntegerDictionaryProvider(createLongDictionaryStreamSources(ImmutableMap.of(nodeId, data)));
+        StreamId streamId = nodeId.toDictionaryDataStreamId();
+        long[] newDictionary = dictionaryProvider.getDictionary(createFlatStreamDescriptor(streamId), buffer, items);
+        assertEquals(newDictionary == buffer, reused);
+        assertTrue(newDictionary.length >= items);
+    }
+
+    @DataProvider(name = "dataForBufferReuseTest")
+    public Object[][] dataForBufferReuseTest()
+    {
+        return new Object[][] {
+                {new long[0], 4, false},
+                {new long[4], 0, true},
+                {new long[4], 4, true},
+                {new long[4], 8, false},
+                {new long[8], 4, true},
+                {new long[16], 12, true},
+                {new long[16], 16, true},
+        };
+    }
+
+    private static class NodeId
+    {
+        private final int node;
+        private final int sequence;
+
+        public NodeId(int node, int sequence)
+        {
+            this.node = node;
+            this.sequence = sequence;
+        }
+
+        public StreamId toDictionaryDataStreamId()
+        {
+            return new StreamId(node, sequence, DICTIONARY_DATA);
+        }
+    }
+
+    private static class DummyOrcDataSource
+            implements OrcDataSource
+    {
+        @Override
+        public OrcDataSourceId getId()
+        {
+            return ORC_DATA_SOURCE_ID;
+        }
+
+        @Override
+        public long getReadBytes()
+        {
+            return 0;
+        }
+
+        @Override
+        public long getReadTimeNanos()
+        {
+            return 0;
+        }
+
+        @Override
+        public long getSize()
+        {
+            return 0;
+        }
+
+        @Override
+        public void readFully(long position, byte[] buffer)
+        {
+        }
+
+        @Override
+        public void readFully(long position, byte[] buffer, int bufferOffset, int bufferLength)
+        {
+        }
+
+        @Override
+        public <K> Map<K, OrcDataSourceInput> readFully(Map<K, DiskRange> diskRanges)
+        {
+            return null;
+        }
+    }
+
+    private StreamDescriptor createFlatStreamDescriptor(StreamId streamId)
+    {
+        return new StreamDescriptor("test_dictionary_stream", streamId.getColumn(),
+                String.format("field_%d", streamId.getColumn()), LONG_TYPE, DUMMY_ORC_DATA_SOURCE, ImmutableList.of(), streamId.getSequence());
+    }
+
+    private InputStreamSources createLongDictionaryStreamSources(Map<NodeId, long[]> streams)
+    {
+        ImmutableMap.Builder<StreamId, InputStreamSource<?>> dictionaryStreamBuilder = ImmutableMap.builder();
```

Comment:
A simple HashMap would work just fine, InputStreamSources uses ImmutableMap.copyOf for its arguments anyway.